### PR TITLE
[Feature][Flink] Rewrite start-seatunnel-flink script with starter

### LINKED ISSUE #1632: [Feature][Flink] Rewrite start-seatunnel-flink script with starter
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/incubator-seatunnel/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

After #1486 has been merged, we make the start-spark-flink script simple.
It's time to make the flink script simple.

I plan to use a FlinkStart to rewrite the start-seatunnel-flink.sh, the advantage is that once we add a parameter of flink task, we don't need to modify the script, we can just modify the code.

### Usage Scenario

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of this pull request

close #1632

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in you PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/development/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by ruanwenjun** [comment]
@simon824 PTAL, thanks.

File: seatunnel-core/seatunnel-core-flink/src/test/java/org/apache/seatunnel/FlinkStarterTest.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class FlinkStarterTest {
+
+    @Test
+    public void buildCommands() {
+        String[] args = {"--config", "test.conf", "-m", "yarn-cluster", "-i", "key1=value1", "-i", "key2=value2"};
+        FlinkStarter flinkStarter = new FlinkStarter(args);
```
Comment (simon824) [inline seatunnel-core/seatunnel-core-flink/src/test/java/org/apache/seatunnel/FlinkStarterTest.java:28]: can we test a negative case, like missing `config` param

Comment (ruanwenjun) [reply seatunnel-core/seatunnel-core-flink/src/test/java/org/apache/seatunnel/FlinkStarterTest.java:28]: Ok, I added a negative case, please review.

**Comment by simon824** [comment]
Overall LGTM, also cc @yx91490

File: seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+     * SeaTunnel parameters, used by SeaTunnel application. e.g. `-c config.conf`
+     */
+    private final FlinkCommandArgs flinkCommandArgs;
+
+    /**
+     * SeaTunnel flink job jar.
+     */
+    private final String appJar;
+
+    public FlinkStarter(String[] args) {
```
Comment (yx91490) [inline seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java]: can be private constructor

File: seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+
+    @Override
+    public List<String> buildCommands() {
+        List<String> command = new ArrayList<>();
+        command.add("${FLINK_HOME}/bin/flink run");
+        command.addAll(flinkParams);
+        command.add("-c " + APP_NAME);
+        command.add(appJar);
+        command.add("--config " + flinkCommandArgs.getConfigFile());
```
Comment (yx91490) [inline seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java]: use '+' maybe make command not atomic, split to 2 add step may be better.

File: seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh [hunk 4]
```diff
@@ -16,87 +16,24 @@
...
 if [ -f "${CONF_DIR}/seatunnel-env.sh" ]; then
-    source ${CONF_DIR}/seatunnel-env.sh
+    . "${CONF_DIR}/seatunnel-env.sh"
 fi
 
-string_trim() {
-    echo $1 | awk '{$1=$1;print}'
-}
-
-JVM_ARGS=$(string_trim "${variables_substitution}")
```
Comment (yx91490) [inline seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh:96]: you changed the JVM_ARGS usage method, I don't known whether will cause problem:(
especially if variables contains quote, blank space, etc.

File: seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java (hunk 2 above)
Comment (ruanwenjun) [reply seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java]: Use public here is just used for test, I removed the public here, and use the default scope.

File: seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java (hunk 3 above)
Comment (ruanwenjun) [reply seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java]: Ok, modified as you suggested.

File: seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh (hunk 4 above)
Comment (ruanwenjun) [reply seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh:96]: Yes, it's better to use trim to process the variables. Done.

File: seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            jCommander.usage();
+            System.exit(USAGE_EXIT_CODE);
+        }
+        return flinkCommandArgs;
+    }
+
+    @Override
+    public List<String> buildCommands() {
+        List<String> command = new ArrayList<>();
+        command.add("${FLINK_HOME}/bin/flink run");
```
Comment (yx91490) [inline seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java]: same

File: seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh (hunk 4 above)
Comment (yx91490) [reply seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh:96]: you may mistake my meaning :(
your code don't set JVM_ARGS environment variable, that's the difference.

Comment (ruanwenjun) [reply seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh:96]: The variables have been set by `-D`

File: seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java (hunk 5 above)
Comment (ruanwenjun) [reply seatunnel-core/seatunnel-core-flink/src/main/java/org/apache/seatunnel/FlinkStarter.java]: Done

File: seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh (hunk 4 above)
Comment (yx91490) [reply seatunnel-core/seatunnel-core-flink/src/main/bin/start-seatunnel-flink.sh:96]: OK

**Comment by yx91490** [comment]
LGTM

**Review by yx91490 — APPROVED** [review]

**Pull request merged**